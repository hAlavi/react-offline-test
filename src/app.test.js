/* eslint-disable react/prop-types */
import React from 'react';
import {
  render,
  cleanup,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import axios from './Helpers/axios';
import App from './app';
import 'jest-styled-components';

jest.mock('./Helpers/axios');

const mockData = {
  data: {
    generationmix: [
      { fuel: 'testFuelMock1', perc: 1 },
      { fuel: 'testFuelMock2', perc: 2 },
    ],
  },
};

afterEach(cleanup);

describe('Unit tests for the App', () => {
  describe('Visual Snapshot testing', () => {
    it('Renders without crashing', () => {
      axios.get.mockImplementation(() => {
        return Promise.resolve({
          status: 200,
          data: mockData,
        });
      });
      const { container } = render(<App />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Functional unit tests', () => {
    it('it applies context from axios - Happy path', async () => {
      axios.get.mockImplementation(() => {
        return Promise.resolve({
          status: 200,
          data: mockData,
        });
      });
      const { getByTestId, getByText } = render(<App />);

      await waitForElementToBeRemoved(() => getByTestId('loading-spinner'));

      expect(getByText('testFuelMock1')).toBeTruthy();
      expect(getByText('testFuelMock2')).toBeTruthy();
    });
    it('it applies context to the consumer, should return Error with Response - Unhappy path', async () => {
      axios.get.mockImplementation(() => {
        // eslint-disable-next-line prefer-promise-reject-errors
        return Promise.reject({
          response: {
            status: 400,
            data: {
              errorMessage: 'ERROR',
            },
          },
        });
      });
      const { getByTestId } = render(<App />);

      await waitForElementToBeRemoved(() => getByTestId('loading-spinner'));

      expect(getByTestId('error-message')).toBeTruthy();
    });

    it('it applies context to the app, should return Error without Response - Unhappy path', async () => {
      axios.get.mockImplementation(() => {
        // eslint-disable-next-line prefer-promise-reject-errors
        return Promise.reject({
          message: 'ERROR',
        });
      });
      const { getByTestId } = render(<App />);

      await waitForElementToBeRemoved(() => getByTestId('loading-spinner'));

      expect(getByTestId('error-message')).toBeTruthy();
    });
  });
});
