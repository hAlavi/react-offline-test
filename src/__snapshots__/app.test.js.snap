// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for the App Visual Snapshot testing Renders without crashing 1`] = `
.c0 {
  position: fixed;
  left: 0;
  top: 0;
  z-index: 99999;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100vw;
  height: 100vh;
  background-color: #999999;
}

.c1 {
  margin: auto;
  font-size: 32px;
  -webkit-animation: spin 2s infinite linear;
  animation: spin 2s infinite linear;
}

<div
  class="c0"
>
  <div
    class="c1"
    data-testid="loading-spinner"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-spinner fa-w-16 "
      data-icon="spinner"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
